{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature performance comparison\n",
    "\n",
    "The purpose of this notebook is to compare the classification performance of the individual features, and their combination, for E. Coli"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from IPython.display import display\n",
    "\n",
    "sys.path.append('../src')\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler\n",
    "from sklearn.svm import SVC, LinearSVC\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.decomposition import PCA, KernelPCA\n",
    "from sklearn.pipeline import make_pipeline, Pipeline\n",
    "from sklearn.feature_selection import SelectKBest, RFE\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.metrics import f1_score, classification_report, confusion_matrix\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict, GridSearchCV\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.base import clone\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from yellowbrick.features import ParallelCoordinates\n",
    "from yellowbrick.features import Rank1D, Rank2D\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "from dataset.transporter_dataset import create_dataset\n",
    "from dataset.cluster_fasta import cd_hit\n",
    "from features.labels import fasta_to_labels\n",
    "from features.compositions import calculate_composition_feature\n",
    "from features.pssm import calculate_pssm_feature\n",
    "from features.coexp import calculate_coexp_feature\n",
    "from models.eval import nested_crossval\n",
    "from visualization.feature_plots import create_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_THREADS = 16\n",
    "IDENTITY_THRESHOLD=70\n",
    "TAX_ID = 9606\n",
    "ORGANISM = \"human\"\n",
    "LOG_FILE = \"../logs/human_amino_sugar.log\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keywords_transport</th>\n",
       "      <th>keywords_location</th>\n",
       "      <th>keywords_transport_related</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9BWM7</th>\n",
       "      <td>Amino-acid transport</td>\n",
       "      <td>Membrane;Mitochondrion;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SFXN3</td>\n",
       "      <td>Sideroflexin-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9606</td>\n",
       "      <td>MGELPLDINIQEPRWDQSTFLGRARHFFTVTDPRNLLLSGAQLEAS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BRV3</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC50A1 RAG1AP1 SCP</td>\n",
       "      <td>Sugar transporter SWEET1 (HsSWEET1) (RAG1-acti...</td>\n",
       "      <td>2.A.123.1.4</td>\n",
       "      <td>9606</td>\n",
       "      <td>MEAGGFLDSLIYGACVVFTLGMFSAGLSDLRHMRMTRSVDNVQFLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q5M8T2</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC35D3 FRCL1</td>\n",
       "      <td>Solute carrier family 35 member D3 (Fringe con...</td>\n",
       "      <td>2.A.7.15.5</td>\n",
       "      <td>9606</td>\n",
       "      <td>MRQLCRGRVLGISVAIAHGVFSGSLNILLKFLISRYQFSFLTLVQC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q969S0</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC35B4 YEA4 PSEC0055</td>\n",
       "      <td>UDP-xylose and UDP-N-acetylglucosamine transpo...</td>\n",
       "      <td>2.A.7.10.2</td>\n",
       "      <td>9606</td>\n",
       "      <td>MRPALAVGLVFAGCCSNVIFLELLARKHPGCGNIVTFAQFLFIAVE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O75387</th>\n",
       "      <td>Amino-acid transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC43A1 LAT3 PB39 POV1</td>\n",
       "      <td>Large neutral amino acids transporter small su...</td>\n",
       "      <td>2.A.1.44.1</td>\n",
       "      <td>9606</td>\n",
       "      <td>MAPTLQQAYRRRWWMACTAVLENLFFSAVLLGWGSLLIILKNEGFY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BYW1</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC2A11 GLUT11</td>\n",
       "      <td>Solute carrier family 2, facilitated glucose t...</td>\n",
       "      <td>2.A.1.1.44</td>\n",
       "      <td>9606</td>\n",
       "      <td>MRALRRLIQGRILLLTICAAGIGGTFQFGYNLSIINAPTLHIQEFT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P11169</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC2A3 GLUT3</td>\n",
       "      <td>Solute carrier family 2, facilitated glucose t...</td>\n",
       "      <td>2.A.1.1.91</td>\n",
       "      <td>9606</td>\n",
       "      <td>MGTQKVTPALIFAITVATIGSFQFGYNTGVINAPEKIIKEFINKTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P14672</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC2A4 GLUT4</td>\n",
       "      <td>Solute carrier family 2, facilitated glucose t...</td>\n",
       "      <td>2.A.1.1.80</td>\n",
       "      <td>9606</td>\n",
       "      <td>MPSGFQQIGSEDGEPPQQRVTGTLVLAVFSAVLGSLQFGYNIGVIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q96AA3</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>RFT1</td>\n",
       "      <td>Protein RFT1 homolog</td>\n",
       "      <td>2.A.66.3.2</td>\n",
       "      <td>9606</td>\n",
       "      <td>MGSQEVLGHAARLASSGLLLQVLFRLITFVLNAFILRFLSKEIVGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q92536</th>\n",
       "      <td>Amino-acid transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Antiport;Transport</td>\n",
       "      <td>SLC7A6 KIAA0245</td>\n",
       "      <td>Y+L amino acid transporter 2 (Cationic amino a...</td>\n",
       "      <td>2.A.3.8.23</td>\n",
       "      <td>9606</td>\n",
       "      <td>MEAREPGRPTPTYHLVPNTSQSQVEEDVSSPPQRSSETMQLKKEIS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           keywords_transport                     keywords_location  \\\n",
       "Uniprot                                                               \n",
       "Q9BWM7   Amino-acid transport  Membrane;Mitochondrion;Transmembrane   \n",
       "Q9BRV3        Sugar transport  Cell membrane;Membrane;Transmembrane   \n",
       "Q5M8T2        Sugar transport                Membrane;Transmembrane   \n",
       "Q969S0        Sugar transport                Membrane;Transmembrane   \n",
       "O75387   Amino-acid transport                Membrane;Transmembrane   \n",
       "...                       ...                                   ...   \n",
       "Q9BYW1        Sugar transport  Cell membrane;Membrane;Transmembrane   \n",
       "P11169        Sugar transport  Cell membrane;Membrane;Transmembrane   \n",
       "P14672        Sugar transport  Cell membrane;Membrane;Transmembrane   \n",
       "Q96AA3        Sugar transport                Membrane;Transmembrane   \n",
       "Q92536   Amino-acid transport  Cell membrane;Membrane;Transmembrane   \n",
       "\n",
       "        keywords_transport_related              gene_names  \\\n",
       "Uniprot                                                      \n",
       "Q9BWM7                   Transport                   SFXN3   \n",
       "Q9BRV3                   Transport     SLC50A1 RAG1AP1 SCP   \n",
       "Q5M8T2                   Transport           SLC35D3 FRCL1   \n",
       "Q969S0                   Transport   SLC35B4 YEA4 PSEC0055   \n",
       "O75387                   Transport  SLC43A1 LAT3 PB39 POV1   \n",
       "...                            ...                     ...   \n",
       "Q9BYW1                   Transport          SLC2A11 GLUT11   \n",
       "P11169                   Transport            SLC2A3 GLUT3   \n",
       "P14672                   Transport            SLC2A4 GLUT4   \n",
       "Q96AA3                   Transport                    RFT1   \n",
       "Q92536          Antiport;Transport         SLC7A6 KIAA0245   \n",
       "\n",
       "                                             protein_names      tcdb_id  \\\n",
       "Uniprot                                                                   \n",
       "Q9BWM7                                      Sideroflexin-3          NaN   \n",
       "Q9BRV3   Sugar transporter SWEET1 (HsSWEET1) (RAG1-acti...  2.A.123.1.4   \n",
       "Q5M8T2   Solute carrier family 35 member D3 (Fringe con...   2.A.7.15.5   \n",
       "Q969S0   UDP-xylose and UDP-N-acetylglucosamine transpo...   2.A.7.10.2   \n",
       "O75387   Large neutral amino acids transporter small su...   2.A.1.44.1   \n",
       "...                                                    ...          ...   \n",
       "Q9BYW1   Solute carrier family 2, facilitated glucose t...   2.A.1.1.44   \n",
       "P11169   Solute carrier family 2, facilitated glucose t...   2.A.1.1.91   \n",
       "P14672   Solute carrier family 2, facilitated glucose t...   2.A.1.1.80   \n",
       "Q96AA3                                Protein RFT1 homolog   2.A.66.3.2   \n",
       "Q92536   Y+L amino acid transporter 2 (Cationic amino a...   2.A.3.8.23   \n",
       "\n",
       "         organism_id                                           sequence  \n",
       "Uniprot                                                                  \n",
       "Q9BWM7          9606  MGELPLDINIQEPRWDQSTFLGRARHFFTVTDPRNLLLSGAQLEAS...  \n",
       "Q9BRV3          9606  MEAGGFLDSLIYGACVVFTLGMFSAGLSDLRHMRMTRSVDNVQFLP...  \n",
       "Q5M8T2          9606  MRQLCRGRVLGISVAIAHGVFSGSLNILLKFLISRYQFSFLTLVQC...  \n",
       "Q969S0          9606  MRPALAVGLVFAGCCSNVIFLELLARKHPGCGNIVTFAQFLFIAVE...  \n",
       "O75387          9606  MAPTLQQAYRRRWWMACTAVLENLFFSAVLLGWGSLLIILKNEGFY...  \n",
       "...              ...                                                ...  \n",
       "Q9BYW1          9606  MRALRRLIQGRILLLTICAAGIGGTFQFGYNLSIINAPTLHIQEFT...  \n",
       "P11169          9606  MGTQKVTPALIFAITVATIGSFQFGYNTGVINAPEKIIKEFINKTL...  \n",
       "P14672          9606  MPSGFQQIGSEDGEPPQQRVTGTLVLAVFSAVLGSLQFGYNIGVIN...  \n",
       "Q96AA3          9606  MGSQEVLGHAARLASSGLLLQVLFRLITFVLNAFILRFLSKEIVGV...  \n",
       "Q92536          9606  MEAREPGRPTPTYHLVPNTSQSQVEEDVSSPPQRSSETMQLKKEIS...  \n",
       "\n",
       "[87 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Delete previous log\n",
    "if os.path.exists(LOG_FILE):\n",
    "    with open(LOG_FILE, \"w\"):\n",
    "        pass\n",
    "\n",
    "create_dataset(\n",
    "    keywords_substrate_filter=[\"Amino-acid transport\", \"Sugar transport\"],\n",
    "    keywords_component_filter=[\"Membrane\"],\n",
    "    keywords_transport_filter=[\"Transport\"],\n",
    "    input_file=\"../data/raw/swissprot/uniprot-reviewed_yes.tab.gz\",\n",
    "    multi_substrate=\"integrate\",\n",
    "    verbose=True,\n",
    "    outliers=[\"Q9HBR0\", \"Q07837\"],\n",
    "    tax_ids_filter=[TAX_ID],\n",
    "    output_tsv=f\"../data/datasets/{ORGANISM}_amino_sugar.tsv\",\n",
    "    output_fasta=f\"../data/datasets/{ORGANISM}_amino_sugar.fasta\",\n",
    "    output_log=LOG_FILE,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd_hit(\n",
    "    executable_location=\"cd-hit\",\n",
    "    input_fasta=f\"../data/datasets/{ORGANISM}_amino_sugar.fasta\",\n",
    "    output_fasta=f\"../data/datasets/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}.fasta\",\n",
    "    log_file=LOG_FILE,\n",
    "    identity_threshold=IDENTITY_THRESHOLD,\n",
    "    n_threads=N_THREADS,\n",
    "    memory=4096,\n",
    "    verbose=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keywords_transport</th>\n",
       "      <th>keywords_location</th>\n",
       "      <th>keywords_transport_related</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9BWM7</th>\n",
       "      <td>Amino-acid transport</td>\n",
       "      <td>Membrane;Mitochondrion;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SFXN3</td>\n",
       "      <td>Sideroflexin-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9606</td>\n",
       "      <td>MGELPLDINIQEPRWDQSTFLGRARHFFTVTDPRNLLLSGAQLEAS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BRV3</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Cell membrane;Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC50A1 RAG1AP1 SCP</td>\n",
       "      <td>Sugar transporter SWEET1 (HsSWEET1) (RAG1-acti...</td>\n",
       "      <td>2.A.123.1.4</td>\n",
       "      <td>9606</td>\n",
       "      <td>MEAGGFLDSLIYGACVVFTLGMFSAGLSDLRHMRMTRSVDNVQFLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q5M8T2</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC35D3 FRCL1</td>\n",
       "      <td>Solute carrier family 35 member D3 (Fringe con...</td>\n",
       "      <td>2.A.7.15.5</td>\n",
       "      <td>9606</td>\n",
       "      <td>MRQLCRGRVLGISVAIAHGVFSGSLNILLKFLISRYQFSFLTLVQC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q969S0</th>\n",
       "      <td>Sugar transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC35B4 YEA4 PSEC0055</td>\n",
       "      <td>UDP-xylose and UDP-N-acetylglucosamine transpo...</td>\n",
       "      <td>2.A.7.10.2</td>\n",
       "      <td>9606</td>\n",
       "      <td>MRPALAVGLVFAGCCSNVIFLELLARKHPGCGNIVTFAQFLFIAVE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O75387</th>\n",
       "      <td>Amino-acid transport</td>\n",
       "      <td>Membrane;Transmembrane</td>\n",
       "      <td>Transport</td>\n",
       "      <td>SLC43A1 LAT3 PB39 POV1</td>\n",
       "      <td>Large neutral amino acids transporter small su...</td>\n",
       "      <td>2.A.1.44.1</td>\n",
       "      <td>9606</td>\n",
       "      <td>MAPTLQQAYRRRWWMACTAVLENLFFSAVLLGWGSLLIILKNEGFY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           keywords_transport                     keywords_location  \\\n",
       "Uniprot                                                               \n",
       "Q9BWM7   Amino-acid transport  Membrane;Mitochondrion;Transmembrane   \n",
       "Q9BRV3        Sugar transport  Cell membrane;Membrane;Transmembrane   \n",
       "Q5M8T2        Sugar transport                Membrane;Transmembrane   \n",
       "Q969S0        Sugar transport                Membrane;Transmembrane   \n",
       "O75387   Amino-acid transport                Membrane;Transmembrane   \n",
       "\n",
       "        keywords_transport_related              gene_names  \\\n",
       "Uniprot                                                      \n",
       "Q9BWM7                   Transport                   SFXN3   \n",
       "Q9BRV3                   Transport     SLC50A1 RAG1AP1 SCP   \n",
       "Q5M8T2                   Transport           SLC35D3 FRCL1   \n",
       "Q969S0                   Transport   SLC35B4 YEA4 PSEC0055   \n",
       "O75387                   Transport  SLC43A1 LAT3 PB39 POV1   \n",
       "\n",
       "                                             protein_names      tcdb_id  \\\n",
       "Uniprot                                                                   \n",
       "Q9BWM7                                      Sideroflexin-3          NaN   \n",
       "Q9BRV3   Sugar transporter SWEET1 (HsSWEET1) (RAG1-acti...  2.A.123.1.4   \n",
       "Q5M8T2   Solute carrier family 35 member D3 (Fringe con...   2.A.7.15.5   \n",
       "Q969S0   UDP-xylose and UDP-N-acetylglucosamine transpo...   2.A.7.10.2   \n",
       "O75387   Large neutral amino acids transporter small su...   2.A.1.44.1   \n",
       "\n",
       "         organism_id                                           sequence  \n",
       "Uniprot                                                                  \n",
       "Q9BWM7          9606  MGELPLDINIQEPRWDQSTFLGRARHFFTVTDPRNLLLSGAQLEAS...  \n",
       "Q9BRV3          9606  MEAGGFLDSLIYGACVVFTLGMFSAGLSDLRHMRMTRSVDNVQFLP...  \n",
       "Q5M8T2          9606  MRQLCRGRVLGISVAIAHGVFSGSLNILLKFLISRYQFSFLTLVQC...  \n",
       "Q969S0          9606  MRPALAVGLVFAGCCSNVIFLELLARKHPGCGNIVTFAQFLFIAVE...  \n",
       "O75387          9606  MAPTLQQAYRRRWWMACTAVLENLFFSAVLLGWGSLLIILKNEGFY...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotations = pd.read_table(f\"../data/datasets/{ORGANISM}_amino_sugar.tsv\", index_col=0)\n",
    "df_annotations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature generation\n",
    "\n",
    "## Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Amino-acid transport    48\n",
       "Sugar transport         34\n",
       "Name: labels, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_to_labels(\n",
    "    input_fasta=f\"../data/datasets/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}.fasta\",\n",
    "    output_tsv=f\"../data/features/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}_labels.tsv\",\n",
    ")\n",
    "df_labels = pd.read_table(\n",
    "    f\"../data/features/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}_labels.tsv\",\n",
    "    index_col=0,\n",
    ")\n",
    "df_labels.labels.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_composition_feature(\n",
    "    input_fasta=f\"../data/datasets/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}.fasta\",\n",
    "    output_tsv=f\"../data/features/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}_aac.tsv\",\n",
    "    feature_type=\"aac\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_aac = pd.read_table(\n",
    "    f\"../data/features/{ORGANISM}_amino_sugar_cluster{IDENTITY_THRESHOLD}_aac.tsv\",\n",
    "    index_col=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_stats(df_features, df_labels_, labels=[\"Amino-acid transport\", \"Sugar transport\"]):\n",
    "    df_stats = pd.concat(\n",
    "        {\n",
    "            \"corr\": df_features.corrwith(\n",
    "                df_labels_.labels.transform(lambda x: 1.0 if x == labels[1] else 0.0)\n",
    "            ),\n",
    "            \"mean\": df_features.mean(),\n",
    "            \"std\": df_features.std(),\n",
    "        },\n",
    "        axis=1,\n",
    "    )\n",
    "\n",
    "    df_stats[\"corr_abs\"] = df_stats[\"corr\"].abs()\n",
    "\n",
    "    df_stats[\"mean0\"] = df_features.loc[df_labels_[df_labels_.labels == labels[0]].index].mean()\n",
    "    df_stats[\"mean1\"] = df_features.loc[df_labels_[df_labels_.labels == labels[1]].index].mean()\n",
    "\n",
    "    df_stats[\"median0\"] = df_features.loc[\n",
    "        df_labels_[df_labels_.labels == labels[0]].index\n",
    "    ].median()\n",
    "    df_stats[\"median1\"] = df_features.loc[\n",
    "        df_labels_[df_labels_.labels == labels[1]].index\n",
    "    ].median()\n",
    "\n",
    "    df_stats[\"mediandiff\"] = (df_stats[\"median0\"] - df_stats[\"median1\"]).abs()\n",
    "    df_stats = df_stats.sort_values(\"mediandiff\", ascending=False)\n",
    "    return df_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_independent_test_set(\n",
    "    df_features, df_labels_, labels=[\"Amino-acid transport\", \"Sugar transport\"], test_size=0.2\n",
    "):\n",
    "    X = df_features.to_numpy()\n",
    "    y = np.where(df_labels_.labels == labels[1], 1, 0)\n",
    "    feature_names = df_features.columns.to_numpy()\n",
    "    sample_names = df_features.index.to_numpy()\n",
    "    (\n",
    "        X_train,\n",
    "        X_test,\n",
    "        y_train,\n",
    "        y_test,\n",
    "        sample_names_train,\n",
    "        sample_names_test,\n",
    "    ) = train_test_split(\n",
    "        X, y, sample_names, stratify=y, random_state=42, shuffle=True, test_size=test_size\n",
    "    )\n",
    "    return (\n",
    "        X_train,\n",
    "        X_test,\n",
    "        y_train,\n",
    "        y_test,\n",
    "        sample_names_train,\n",
    "        sample_names_test,\n",
    "        feature_names,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_validation_results(y_true_, y_pred_, labels = [\"Amino\", \"Sugar\"]):\n",
    "    report_dict = classification_report(y_true=y_true_, y_pred=y_pred_, output_dict=True)\n",
    "    report_dict = {\n",
    "        labels[0]: report_dict['0'],\n",
    "        labels[1]: report_dict['1'],\n",
    "        \"Macro\": report_dict[\"macro avg\"],\n",
    "        \"Weighted\": report_dict[\"weighted avg\"]\n",
    "    }\n",
    "    report_df = pd.DataFrame.from_dict(report_dict)\n",
    "    confusion_matrix_df = pd.DataFrame(\n",
    "        confusion_matrix(y_true_, y_pred_),\n",
    "        columns=labels,\n",
    "        index=labels,\n",
    "    )\n",
    "    return report_df, confusion_matrix_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual Features\n",
    "\n",
    "## AAC\n",
    "\n",
    "### Stats, Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corr</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>corr_abs</th>\n",
       "      <th>mean0</th>\n",
       "      <th>mean1</th>\n",
       "      <th>median0</th>\n",
       "      <th>median1</th>\n",
       "      <th>mediandiff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.330764</td>\n",
       "      <td>0.078460</td>\n",
       "      <td>0.016179</td>\n",
       "      <td>0.330764</td>\n",
       "      <td>0.073983</td>\n",
       "      <td>0.084779</td>\n",
       "      <td>0.076141</td>\n",
       "      <td>0.084391</td>\n",
       "      <td>0.008250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>-0.243412</td>\n",
       "      <td>0.067497</td>\n",
       "      <td>0.019474</td>\n",
       "      <td>0.243412</td>\n",
       "      <td>0.071462</td>\n",
       "      <td>0.061899</td>\n",
       "      <td>0.072837</td>\n",
       "      <td>0.064867</td>\n",
       "      <td>0.007970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>-0.185978</td>\n",
       "      <td>0.047826</td>\n",
       "      <td>0.012097</td>\n",
       "      <td>0.185978</td>\n",
       "      <td>0.049708</td>\n",
       "      <td>0.045169</td>\n",
       "      <td>0.049634</td>\n",
       "      <td>0.041925</td>\n",
       "      <td>0.007709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.264524</td>\n",
       "      <td>0.133139</td>\n",
       "      <td>0.023298</td>\n",
       "      <td>0.264524</td>\n",
       "      <td>0.127984</td>\n",
       "      <td>0.140417</td>\n",
       "      <td>0.127623</td>\n",
       "      <td>0.134506</td>\n",
       "      <td>0.006883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>-0.204271</td>\n",
       "      <td>0.032058</td>\n",
       "      <td>0.010375</td>\n",
       "      <td>0.204271</td>\n",
       "      <td>0.033831</td>\n",
       "      <td>0.029555</td>\n",
       "      <td>0.033461</td>\n",
       "      <td>0.027875</td>\n",
       "      <td>0.005586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.129789</td>\n",
       "      <td>0.081882</td>\n",
       "      <td>0.019591</td>\n",
       "      <td>0.129789</td>\n",
       "      <td>0.079755</td>\n",
       "      <td>0.084885</td>\n",
       "      <td>0.076779</td>\n",
       "      <td>0.081950</td>\n",
       "      <td>0.005170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>-0.072430</td>\n",
       "      <td>0.035918</td>\n",
       "      <td>0.012764</td>\n",
       "      <td>0.072430</td>\n",
       "      <td>0.036691</td>\n",
       "      <td>0.034826</td>\n",
       "      <td>0.038704</td>\n",
       "      <td>0.033984</td>\n",
       "      <td>0.004720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>-0.193269</td>\n",
       "      <td>0.026965</td>\n",
       "      <td>0.010087</td>\n",
       "      <td>0.193269</td>\n",
       "      <td>0.028596</td>\n",
       "      <td>0.024663</td>\n",
       "      <td>0.028117</td>\n",
       "      <td>0.023889</td>\n",
       "      <td>0.004228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.126909</td>\n",
       "      <td>0.037150</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.126909</td>\n",
       "      <td>0.035840</td>\n",
       "      <td>0.039001</td>\n",
       "      <td>0.033205</td>\n",
       "      <td>0.037126</td>\n",
       "      <td>0.003920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>-0.113785</td>\n",
       "      <td>0.080466</td>\n",
       "      <td>0.012851</td>\n",
       "      <td>0.113785</td>\n",
       "      <td>0.081689</td>\n",
       "      <td>0.078739</td>\n",
       "      <td>0.081725</td>\n",
       "      <td>0.077942</td>\n",
       "      <td>0.003783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.045365</td>\n",
       "      <td>0.030868</td>\n",
       "      <td>0.009100</td>\n",
       "      <td>0.045365</td>\n",
       "      <td>0.030522</td>\n",
       "      <td>0.031355</td>\n",
       "      <td>0.027631</td>\n",
       "      <td>0.031304</td>\n",
       "      <td>0.003673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.035869</td>\n",
       "      <td>0.059885</td>\n",
       "      <td>0.012966</td>\n",
       "      <td>0.035869</td>\n",
       "      <td>0.059496</td>\n",
       "      <td>0.060434</td>\n",
       "      <td>0.060662</td>\n",
       "      <td>0.057680</td>\n",
       "      <td>0.002982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-0.083771</td>\n",
       "      <td>0.019328</td>\n",
       "      <td>0.007002</td>\n",
       "      <td>0.083771</td>\n",
       "      <td>0.019818</td>\n",
       "      <td>0.018635</td>\n",
       "      <td>0.021333</td>\n",
       "      <td>0.018484</td>\n",
       "      <td>0.002849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>-0.168546</td>\n",
       "      <td>0.028725</td>\n",
       "      <td>0.008551</td>\n",
       "      <td>0.168546</td>\n",
       "      <td>0.029931</td>\n",
       "      <td>0.027023</td>\n",
       "      <td>0.029721</td>\n",
       "      <td>0.027311</td>\n",
       "      <td>0.002410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-0.103455</td>\n",
       "      <td>0.040860</td>\n",
       "      <td>0.011870</td>\n",
       "      <td>0.103455</td>\n",
       "      <td>0.041887</td>\n",
       "      <td>0.039410</td>\n",
       "      <td>0.041368</td>\n",
       "      <td>0.038978</td>\n",
       "      <td>0.002391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>-0.032227</td>\n",
       "      <td>0.034606</td>\n",
       "      <td>0.009848</td>\n",
       "      <td>0.032227</td>\n",
       "      <td>0.034871</td>\n",
       "      <td>0.034231</td>\n",
       "      <td>0.036542</td>\n",
       "      <td>0.034229</td>\n",
       "      <td>0.002313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.098590</td>\n",
       "      <td>0.076765</td>\n",
       "      <td>0.014761</td>\n",
       "      <td>0.098590</td>\n",
       "      <td>0.075548</td>\n",
       "      <td>0.078484</td>\n",
       "      <td>0.076339</td>\n",
       "      <td>0.074184</td>\n",
       "      <td>0.002155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-0.008194</td>\n",
       "      <td>0.056655</td>\n",
       "      <td>0.010684</td>\n",
       "      <td>0.008194</td>\n",
       "      <td>0.056728</td>\n",
       "      <td>0.056551</td>\n",
       "      <td>0.056305</td>\n",
       "      <td>0.054466</td>\n",
       "      <td>0.001839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>-0.024240</td>\n",
       "      <td>0.016931</td>\n",
       "      <td>0.007565</td>\n",
       "      <td>0.024240</td>\n",
       "      <td>0.017084</td>\n",
       "      <td>0.016714</td>\n",
       "      <td>0.016071</td>\n",
       "      <td>0.015086</td>\n",
       "      <td>0.000985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>-0.108182</td>\n",
       "      <td>0.014017</td>\n",
       "      <td>0.006174</td>\n",
       "      <td>0.108182</td>\n",
       "      <td>0.014575</td>\n",
       "      <td>0.013228</td>\n",
       "      <td>0.013471</td>\n",
       "      <td>0.013346</td>\n",
       "      <td>0.000125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       corr      mean       std  corr_abs     mean0     mean1   median0  \\\n",
       "G  0.330764  0.078460  0.016179  0.330764  0.073983  0.084779  0.076141   \n",
       "I -0.243412  0.067497  0.019474  0.243412  0.071462  0.061899  0.072837   \n",
       "P -0.185978  0.047826  0.012097  0.185978  0.049708  0.045169  0.049634   \n",
       "L  0.264524  0.133139  0.023298  0.264524  0.127984  0.140417  0.127623   \n",
       "N -0.204271  0.032058  0.010375  0.204271  0.033831  0.029555  0.033461   \n",
       "A  0.129789  0.081882  0.019591  0.129789  0.079755  0.084885  0.076779   \n",
       "K -0.072430  0.035918  0.012764  0.072430  0.036691  0.034826  0.038704   \n",
       "D -0.193269  0.026965  0.010087  0.193269  0.028596  0.024663  0.028117   \n",
       "R  0.126909  0.037150  0.012346  0.126909  0.035840  0.039001  0.033205   \n",
       "V -0.113785  0.080466  0.012851  0.113785  0.081689  0.078739  0.081725   \n",
       "Q  0.045365  0.030868  0.009100  0.045365  0.030522  0.031355  0.027631   \n",
       "F  0.035869  0.059885  0.012966  0.035869  0.059496  0.060434  0.060662   \n",
       "C -0.083771  0.019328  0.007002  0.083771  0.019818  0.018635  0.021333   \n",
       "M -0.168546  0.028725  0.008551  0.168546  0.029931  0.027023  0.029721   \n",
       "E -0.103455  0.040860  0.011870  0.103455  0.041887  0.039410  0.041368   \n",
       "Y -0.032227  0.034606  0.009848  0.032227  0.034871  0.034231  0.036542   \n",
       "S  0.098590  0.076765  0.014761  0.098590  0.075548  0.078484  0.076339   \n",
       "T -0.008194  0.056655  0.010684  0.008194  0.056728  0.056551  0.056305   \n",
       "W -0.024240  0.016931  0.007565  0.024240  0.017084  0.016714  0.016071   \n",
       "H -0.108182  0.014017  0.006174  0.108182  0.014575  0.013228  0.013471   \n",
       "\n",
       "    median1  mediandiff  \n",
       "G  0.084391    0.008250  \n",
       "I  0.064867    0.007970  \n",
       "P  0.041925    0.007709  \n",
       "L  0.134506    0.006883  \n",
       "N  0.027875    0.005586  \n",
       "A  0.081950    0.005170  \n",
       "K  0.033984    0.004720  \n",
       "D  0.023889    0.004228  \n",
       "R  0.037126    0.003920  \n",
       "V  0.077942    0.003783  \n",
       "Q  0.031304    0.003673  \n",
       "F  0.057680    0.002982  \n",
       "C  0.018484    0.002849  \n",
       "M  0.027311    0.002410  \n",
       "E  0.038978    0.002391  \n",
       "Y  0.034229    0.002313  \n",
       "S  0.074184    0.002155  \n",
       "T  0.054466    0.001839  \n",
       "W  0.015086    0.000985  \n",
       "H  0.013346    0.000125  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_stats(df_aac, df_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAd0AAAFaCAYAAABbiIcOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAkk0lEQVR4nO3deVRV9f7/8dcBRMQBxdKcaimJkpagKLlI/UaiJSJq2tXl1a5pTqm3m3bLtDRLM7tlNyPNBs1uw3UgyanBVDSXEwqaFcmlbFIhlFBRQOD8/ujnSXLgHNx709k+H2v5xxn3+7wjXnz23me/HU6n0ykAAGA6n6ouAACAqwWhCwCARQhdAAAsQugCAGARQhcAAIsQugAAWITQRZVLSkrS4MGDTXnvJ554QomJiaa8t5F27typrl27XvSx1NRU9ezZ0+KKJKfTqSlTpqhjx44aMGCA5dsH7IjQhSVSU1M1aNAgdejQQZ06ddKgQYO0f/9+07c7c+ZMPfDAA4a8V1JSksLCwhQREaH27durT58+2rRpkyHvfTmRkZH6+OOPTd/OH+3Zs0fbtm1TSkqKVqxYccHjmzdv1uDBgxUZGano6GhNmzZNp06dcj1eXFysKVOmqH379oqOjtbixYsvua2dO3eqdevWioiIcP0bM2bMFdV/uT9kgKriV9UFwP5OnTqlMWPGaMaMGbrrrrt09uxZpaamyt/fv0rrKikpkZ+fZ/8LhIeH67333lNZWZmWLVumhx56SCkpKapTp45JVVadn3/+WU2aNFFgYOBFHz958qTGjh2rjh07qri4WJMmTdLcuXM1c+ZMSdL8+fP1/fffa9OmTcrNzdWwYcMUEhJyySBs0KCBtmzZYtrn8VRlfj6AirDShem+++47SVLv3r3l6+urgIAA3XbbbWrdunW55z377LPq2LGjYmJilJKS4rp/5cqVuuuuuxQREaE77rhD77//vuuxc6uZhQsXKioqSjExMfrwww9djz/66KOaN29euecuWrRI0dHRmjJlioqLizVr1izddtttuu222zRr1iwVFxdX+Jl8fHyUkJCg06dP69ChQ5KkH374QcOGDVNUVJSioqI0adIknThxwvWamJgYvfHGG4qPj1eHDh304IMPqqio6KLvv3TpUvXq1UtHjx69YMVW0fu89tprrs+zfPlytWrVSt9///1Ft5Odna0xY8aoU6dOio2N1bJlyyRJy5cv17Rp05Senq6IiAi99NJLF7w2Pj5eXbt2VY0aNRQUFKR77rlHaWlprsdXrVqlcePGKSgoSCEhIRo4cKA++OCDCnv7R+np6Ro0aJAiIyPVp08f7dy50/XYpX42Tp8+rfvvv185OTmulXN2dna5nwfpwtVwTEyMFi1apPj4eIWHh6ukpOSy209KStIdd9yhiIiIC372gIvhzziYrnnz5vL19dUjjzyiXr16KTw8XEFBQeWes3//fvXr1087duzQf//7X02dOlVbt26Vw+FQ/fr19eqrr6pZs2bavXu37r//ft18881q06aNJCk3N1d5eXnaunWr0tPTNWrUKLVt21YtWrS4oJbc3Fzl5+dr06ZNKisr04IFC7Rv3z4lJyfL4XBo3LhxeuWVV/Tggw9e9jOVlpYqKSlJ1apVU5MmTST9dgx09OjR6tixo06dOqUJEyZo/vz5mjp1qut169ev1+uvv67q1atr8ODBFz2enZiYqA0bNug///mPgoODLxqYl3qfLVu2aMmSJVqyZImaNm2qJ5544rKfY9KkSbrxxhu1detWffvttxo+fLiaNWumgQMHytfXV8uXL9d777132fc4Z/fu3brxxhslSfn5+crJySn3h1Xr1q312WefufVe52RnZ2v06NGaO3euunTpou3bt2vixIlav369goODL/uz8dprr+nhhx/2ePW8du1aLVq0SPXq1dOxY8cuuf2AgAA9/fTTWrFihVq0aKGcnBzl5+d7tC1cfVjpwnS1atXSu+++K4fDoccff1ydO3fWmDFjlJub63pO48aNdc8998jX11f9+vXTL7/84nr8//7v/3T99dfL4XCoU6dOio6OVmpqarlt/P3vf5e/v786deqkbt26af369RetxcfHRxMnTpS/v78CAgK0evVqPfDAA6pfv76Cg4P1wAMPXHa1sm/fPkVGRuqWW27Rs88+q7lz56p+/fqSpBtuuEHR0dHy9/dXcHCwhg8frt27d5d7/dChQ9WwYUPVrVtXt99+u77++mvXY06nU88884w+//xzLV26VMHBwZes41Lvs379evXv318tW7ZUjRo1NH78+Eu+x5EjR7Rnzx5NnjxZ1atXV1hYmAYOHKjk5ORLvuZStm3bplWrVmnixImSfltpSlLt2rVdz6ldu7YKCgou+R45OTmKjIx0/Vu3bp2Sk5PVtWtXdevWTT4+PoqOjlbbtm1de0Lc+dnw1NChQ9WoUSMFBARUuH0fHx9lZmaqsLBQDRo0UMuWLa9o27A/VrqwREhIiObMmSNJysrK0sMPP6zZs2frhRdekCRdc801rufWqFFD0u+/uFNSUpSYmKhDhw6prKxMhYWFCg0NdT2/Tp065Y47Nm7cWDk5OReto169eqpevbrrdk5Ojho3buzWayWpXbt2eu+991RQUKCpU6dqz5496tWrlyTp2LFjevrpp5WamqqCggI5nc4LjvVee+215T7n+ds6efKkli1bpnnz5pULq4u51Pvk5OSobdu2rscaNWp0yffIyclRUFCQatWqVe7zHzhw4LLb/qP09HRNmjRJL730kpo3by5Jrv8ep06dcvX71KlTqlmz5iXf52LHdGfMmKGPPvqo3AlrJSUlioqKklTxz0ZlnN+zw4cPX3L7gYGBmjdvnt58801NnTpV7du31yOPPKKQkJAr2j7sjZUuLBcSEqL+/fsrMzOzwucWFxdr4sSJuu+++7Rt2zalpqaqa9euOn841okTJ1wBLf22gmvQoMFF38/hcJS73aBBAx0+fNit156vZs2amj59upKTk/XVV19Jkp5//nk5HA59+OGH2rt3r5577jl5MsSrTp06WrhwoaZMmaI9e/a4/brzNWjQQNnZ2a7bR44cuexz8/Pzy51xfOTIETVs2NDt7X311VcaO3asZs+erc6dO7vuDwoK0rXXXquMjAzXfRkZGa7dz+5q1KiREhISlJqa6vp37hBCRT8bf/xvLf32B0phYaHr9vl7W845/3WX274kdenSRYsXL9bnn3+uFi1a6PHHH/fo8+HqQ+jCdFlZWXrzzTd19OhRSb/9Yl+zZo3atWtX4WuLi4tVXFys4OBg+fn5KSUlRdu2bbvgefPnz1dxcbFSU1O1efNm3XnnnW7VFhcXpwULFuj48eM6fvy4EhMTFR8f79Zr69Wrp4EDB7q+B1xQUKDAwEDVqVNH2dnZev311916n/NFRUXpX//6l8aPH699+/Z5/Po777xTSUlJysrK0pkzZy77HeVGjRopIiJCL7zwgoqKipSRkaEVK1a4/fkPHjyokSNH6vHHH1dMTMwFj/ft21cLFixQfn6+srKytHz5cvXr18+jz3Pua1lbt25VaWmpioqKtHPnTh09erTCn4369evr119/1cmTJ133hYWFKSUlRb/++qt++eUXvfXWW5Xefm5urj777DOdPn1a/v7+CgwMlK+vr0efD1cfQhemq1Wrlvbt26eBAwcqPDxc99xzj0JDQ/Xoo4+69dpp06bpwQcfVMeOHbVmzZoLfsFfc801qlOnjrp06aLJkydrxowZbu/iGzdunNq2bas+ffqoT58+atOmjcaNG+f2Z7v33nuVkpKijIwMjR8/Xl999ZUiIyM1atQo9ejRw+33OV90dLSeeeYZjR071uNdvd26ddPQoUM1bNgwxcbGKjw8XJIu+fWsF154QT///LO6dOmi8ePHa8KECYqOjnZrW4sXL9bx48c1depU1xnCcXFxrscnTpyoZs2a6fbbb9fQoUM1YsQIj78326hRI73yyit69dVX1blzZ3Xr1k1vvPGGysrKKvzZCAkJUVxcnLp3767IyEhlZ2crISFBrVu3VkxMjO677z7XoYHKbL+srEyLFy9Wly5d1KlTJ+3evVvTp0/36PPh6uNgiD282c6dOyt1hurVIisrS71799YXX3zBd06BPwFWuoDNfPrppyouLlZ+fr6ee+453X777QQu8CdB6AI28/7776tz586KjY2Vr6+vZsyYUdUlAfj/2L0MAIBFWOkCAGARUw/0lJWVqaCgQNWqVbvod+YAALATp9Ops2fPqmbNmvLxuXBda2roFhQU6ODBg2ZuAgCAP53Q0NCLXlnO1NCtVq2aa+NVPcbNbg4cOFDucn+4cvTUHPTVHPTVeEb0tLi4WAcPHnTl3x+ZGrrndin7+/uXu94tjEFPjUdPzUFfzUFfjWdUTy91SJUTqQAAsAihCwCARQhdAAAsQugCAGARQhcAAIsQugAAWITQBQDAIoQuAAAWIXQBALAIoQsAgEVMvQzkOSGzPtCRgrNWbOrq8u5XVV2B/dBTc9BXc9DXCpU+P7SqSyiHlS4AABbxOHQ//fRTtWrVSllZWWbUAwCAbXkcumvWrFGHDh20bt06M+oBAMC2PArdgoIC7d27V7NmzdLatWvNqgkAAFvyKHQ3bNigLl26qHnz5qpbt66+/PJLs+oCAMB2PArdtWvXKi4uTpLUq1cvrVmzxpSiAACwI7e/MpSXl6cdO3YoMzNTDodDpaWlcjgc+uc//ymHw2FmjQAA2ILbK92PP/5Yffv21aZNm7Rx40alpKSoadOm2rNnj5n1AQBgG26H7tq1a9W9e/dy9/Xo0UOrV682vCgAAOzI7d3Lb7/99gX3DRs2zNBiAACwM0suA5k1tZ+qV69uxaauGnv27FGHDh2qugxboafmoK/moK/eictAAgBgEUIXAACLMGXImzFhxHj01Bz01Rz09YpZPYWIlS4AABbxeKX7yy+/aPbs2friiy/k7++vJk2a6LHHHlPz5s3NqA8AANvwKHSdTqfGjx+vvn37at68eZKkr7/+WseOHSN0AQCogEehu2PHDvn5+Wnw4MGu+8LCwgwvCgAAO/LomG5mZqbatGljVi0AANgaJ1IBAGARj0K3ZcuWzNAFAKCSPArdW2+9VcXFxVq2bJnrvv3792vXrl2GFwYAgN14FLoOh0Mvv/yytm3bpu7duysuLk4vv/yyGjRoYFZ9AADYhsff023YsKH+/e9/m1ELAAC2xpQhL8WEEePRU3PQV3PQV+/E2csAAFiE0AUAwCJMGfJmTBgxHj01B301x1XQV6unAJmNlS4AABZxe6UbFham0NBQlZSUyNfXV/369dO9994rHx9yGwAAd7gdugEBAUpOTpYkHTt2TJMmTdLJkyc1ceJE04oDAMBOKrVMrV+/vp566im98847cjqdRtcEAIAtVXrfcLNmzVRWVqZjx44ZWQ8AALZ1RQdkWeUCAOC+Sofujz/+KF9fX9WvX9/IegAAsK1Khe7x48c1ffp0DRkyRA6Hw+iaAACwJbfPXi4sLFRCQoLrK0MJCQkaPny4mbUBAGArbofu119/bWYdAADYHlOGvBQTRoxHT81BX81BX70Tl5MCAMAihC4AABZhypA3uwomjFiOnpqDvpqDvpbjDROJPA7dc4MPzomLi9OoUaMMLQoAADvyOHTPH3wAAADcxzFdAAAs4vFK99xFMs4ZPXq0evXqZWhRAADYEbuXAQCwCLuXAQCwCKELAIBFrviYbpcuXTR58mRDiwIAwI48Dl0GHwAAUDnsXgYAwCJMGfJSTBgxHj01B301B331Tqx0AQCwCAMPvBkXOzcePTUHfTUHfb0iVTEggZUuAAAWueIpQ4mJiWratKmhRQEAYEdcBhIAAIuwexkAAItc0RWpmjZtqsTERMOLAgDAjti9DACARdi9DACARQhdAAAsQugCAGARj0M3LS3NjDoAALA9Bh54KS52bjx6ag76ag766p3YvQwAgEUIXQAALMKUIW/GhBHj0VNz0Fdz0NcrwpQhAABszKOVbm5urp555hmlp6crKChI1apV08iRIxUbG2tWfQAA2Ibboet0OvXAAw+ob9++ev755yVJP//8szZu3GhacQAA2Inbu5d37NihatWqafDgwa77mjRpoqFDrd8nDgCAN3I7dDMzM3XTTTeZWQsAALZW6bOXn3zySe3Zs0fVqlXTypUrjawJAABbcnul27JlS3311e+np0+fPl1LlixRXl6eKYUBAGA3bofurbfeqqKiIr377ruu+woLC00pCgAAO3J797LD4VBiYqKeeeYZvf766woODlaNGjU0efJkM+sDAMA2PDqm26BBA82bN8+sWgAAsDWmDHkpJowYj56ag76ag756Jy4DCQCARQhdAAAswpQhb8aEEePRU3PQV3PQ10qriglDEitdAAAs41HoRkRElLudlJSkmTNnGloQAAB2xUoXAACLELoAAFjEoxOpCgsLlZCQ4Lqdn5+vmJgYw4sCAMCOPArdgIAAJScnu24nJSXpwIEDhhcFAIAdsXsZAACLELoAAFiE0AUAwCIeHdNNS0srd7t///7q37+/oQUBAGBXTBnyUkwYMR49NQd9NQd99U7sXgYAwCKELgAAFmHKkDdjwojx6Kk56Ks56GuleeWUoT8OQAAAAJfG7mUAACxC6AIAYBFCFwAAixC6AABYhNAFAMAihC4AABYhdAEAsMgVhe4fByAAAIBLY+CBl+Ji58ajp+agr+agr96J3csAAFiE0AUAwCIMPPBmXOzcePTUHPTVHPTVI1U15OB8rHQBALCIx6F7/mShlJQU9ejRQ4cPHza0KAAA7KjSu5e3b9+up556Sm+++aYaN25sZE0AANhSpUI3NTVV06ZN06JFi3T99dcbXRMAALbkcegWFxdr3LhxWrp0qUJCQsyoCQAAW/L4mK6fn58iIiK0YsUKM+oBAMC2PA5dHx8fvfjiizpw4IAWLlxoRk0AANhSpb4yVKNGDS1cuFCrV6/W8uXLja4JAABbqvT3dOvWravXX39dCxYs0IYNG4ysCQAAW/L4RKrzJws1atRIGzduNLQgAADsiilDXooJI8ajp+agr+agr96Jy0ACAGARQhcAAIswZcibMWHEePTUHPTVHPS1nD/DFKGKsNIFAMAilQrd8ycNAQAA97DSBQDAIoQuAAAWIXQBALAIoQsAgEUIXQAALELoAgBgkUpdHOPMmTPq2rWr6/bw4cM1fPhww4oCAMCOKhW6GRkZRtcBAIDtMWXISzFhxHj01Bz01Rz01TtxTBcAAIsQugAAWIQpQ96MCSPGo6fmoK/moK9u+TNNH2KlCwCARdwK3VatWunhhx923S4pKdGtt96q0aNHm1YYAAB241boBgYGKjMzU4WFhZKkbdu2qWHDhqYWBgCA3bi9e7lr167avHmzJGnt2rWKi4szqyYAAGzJ7dDt1auX1q1bp6KiIn3zzTdq166dmXUBAGA7bodu69at9dNPP2nNmjXq1q2bmTUBAGBLHp29HBMTo7lz57JrGQCASvDoe7oDBgxQ7dq11apVK+3cudOsmgAAsCWPVrrXXXed7r33XrNqAQDA1txa6aalpV1wX1RUlKKiogwvCAAAu2LKkJdiwojx6Kk56Ks56Kt34jKQAABYhIEH3oyLnRuPnpqDvpqDvl4xq4chsNIFAMAiHoVuq1atNGfOHNftN954Q/Pnzze8KAAA7Mij0PX399cnn3yi48ePm1UPAAC25VHo+vn56S9/+Yveeusts+oBAMC2PD6mO2TIEK1evVonT540ox4AAGzL49CtVauWEhIStHTpUjPqAQDAtip19vK9996rlStX6syZM0bXAwCAbVUqdOvWras777xTK1asMLoeAABsq9Lf073vvvuUl5dnZC0AANiaR1ekOn/wwTXXXKN9+/YZXhAAAHbFwAMvxcXOjUdPzUFfzUFfvROXgQQAwCKELgAAFmHKkDdjwojx6Kk56Ks56GulWT1d6BxWugAAWMTjlW5YWJhCQ0NVWlqqFi1a6Nlnn1WNGjXMqA0AAFvxeKUbEBCg5ORkrVmzRtWqVdP7779vRl0AANjOFe1ejoyM1Pfff29ULQAA2FqlQ7ekpERbtmxRaGiokfUAAGBbHh/TLSwsVEJCgqTfVroDBgwwvCgAAOzI49A9d0wXAAB4hq8MAQBgEUIXAACLeBy6508aAgAA7mPKkJdiwojx6Kk56Ks56Kt3YvcyAAAWIXQBALAIU4a8GRNGjEdPzUFfzUFfL6qqJgi5g5UuAAAW8Wile/ToUT355JPKyspSaWmpunXrpkcffVT+/v5m1QcAgG24vdJ1Op0aP368unfvrk8++USffPKJCgsLNXfuXDPrAwDANtwO3R07dqh69eq6++67JUm+vr567LHHlJycrIKCAtMKBADALtwO3czMTLVp06bcfbVq1VKTJk30ww8/GF4YAAB249HuZYfDcdH7AQBAxdwO3ZYtW+rAgQPl7jt16pSOHTum5s2bG14YAAB243bodu7cWWfOnNGqVaskSaWlpZozZ46GDBmigIAAs+oDAMA23A5dh8OhxMREffTRR+rRo4eioqLk4+OjsWPHmlkfAAC24dH3dBs1aqSFCxdKkvbu3atJkybpwIEDatu2rSnFAQBgJ5W+DGT79u21adMmt57LlCHjMWHEePTUHPTVHPTVO3EZSAAALELoAgBgEaYMeTMmjBiPnpqDvprjKu/rn3ma0KV4FLphYWEKDQ1VaWmpmjZtqrlz56pOnTpm1QYAgK14tHs5ICBAycnJWrNmjYKCgvTOO++YVRcAALZT6WO64eHhys7ONrIWAABsrVKhW1paqu3btysmJsboegAAsC2PQrewsFAJCQmKiopSfn6+oqOjzaoLAADbqdQx3U2bNuns2bMc0wUAwAOV2r1cu3ZtTZs2TW+++abOnuWrQAAAuKPSJ1LddNNNat26tdauXWtkPQAA2JZH39NNS0srd/vc8AMAAFAxLgMJAIBFLLkMJFOGjMeEEePRU3PQV3PQV+/EShcAAIsw8MCbXeUXOzcFPTUHfTUHfa20qhqWwEoXAACLeLzSXbBggdasWSMfHx/5+Pho5syZateunRm1AQBgKx5/ZWjz5s364IMP5O/vr+PHj3NxDAAA3ORR6P7yyy+qV6+e/P39JUnBwcGmFAUAgB15dEw3OjpaR44cUc+ePTVjxgzt2rXLrLoAALAdj0K3Zs2aSkpK0syZMxUcHKx//OMfSkpKMqs2AABsxeMTqXx9fRUVFaWoqCiFhoZq1apV6t+/vxm1AQBgKx6tdL/99lsdOnTIdfvrr79W48aNja4JAABb8mile/r0aT399NM6ceKEfH19dcMNN2jmzJlm1QYAgK14FLpt27bV+++/b1YtAADYGgMPvBQXOzcePTUHfTUHffVOXAYSAACLELoAAFiEKUPejAkjxqOn5qCv5qCvlVJVE4YkVroAAFjGo5VuXl6e/va3v0mScnNz5ePj47r+8vLly13XZAYAABfyKHTr1aun5ORkSdL8+fMVGBioESNGmFIYAAB2w+5lAAAsQugCAGARQhcAAIsQugAAWITQBQDAIoQuAAAWqfQVqSZMmGBkHQAA2B5ThrwUE0aMR0/NQV/NQV+9E7uXAQCwCKELAIBFmDLkzZgwYjx6ag76ag76esWsnjjEShcAAIu4HbpDhw7V1q1by923ZMkSzZgxw+iaAACwJbdDNy4uTuvWrSt337p169S7d2/DiwIAwI7cDt2ePXtq06ZNKi4uliT99NNPysnJ4ZR1AADc5Hbo1qtXT7fccou2bNki6bdV7l133SWHw2FacQAA2IlHJ1Kdv4t57dq17FoGAMADHoVu9+7dtX37dn355ZcqLCxUmzZtzKoLAADb8Sh0a9asqU6dOumxxx5jlQsAgIc8/p5u7969lZGRoV69eplRDwAAtuXxFaliY2P1zTffmFELAAC2xpQhL8WEEePRU3PQV3PQV+/EZSABALAIoQsAgEWYMuTNmDBiPHpqDvpqDvpaIaunCFWkwpXu7NmztWTJEtftESNGaOrUqa7bc+bM0eLFi00pDgAAO6kwdCMiIpSWliZJKisrU15env73v/+5Hk9LS1P79u3NqxAAAJuoMHTbt2/vCt3MzEy1bNlSNWvWVH5+voqLi5WVlaWwsDDTCwUAwNtVeEy3YcOG8vPz0+HDh5WWlqbw8HBlZ2crPT1dtWrVUqtWreTv729FrQAAeDW3TqQ6t4s5LS1Nw4cPV3Z2tvbu3avatWsrIiLC7BoBALAFt74ydG4X88GDB9WyZUu1a9dO6enpHM8FAMADbofupk2bFBQUJF9fX9WtW1cnT55Uenq6wsPDTS4RAAB7cCt0Q0NDlZeXp3bt2pW7r1atWgoODjatOAAA7MStY7q+vr7au3dvufvmzJljSkEAANgVAw+8FBc7Nx49NQd9NQd99U5cexkAAIsQugAAWISBB96Mi50bj56ag76ag75WWlUNQmClCwCARTwKXafTqcGDByslJcV137p16zRixAjDCwMAwG48Cl2Hw6Enn3xSc+bMUVFRkU6fPq0XX3xR06dPN6s+AABsw+NjuqGhobr99tv12muv6fTp00pISND1119vRm0AANhKpU6kGj9+vPr16yd/f3+tXLnS6JoAALClSoVuYGCgevXqpcDAQMb6AQDgpkqfvezj4yMfH05+BgDAXaQmAAAWIXQBALBIpa9INWHCBCPrAADA9pgy5KWYMGI8emoO+moO+uqd2L0MAIBFCF0AACxC6AIAYBFCFwAAixC6AABYhNAFAMAihC4AABYhdAEAsAihCwCARQhdAAAsQugCAGARQhcAAIuYOvDA6XRKkoqLi83czFWrqKioqkuwHXpqDvpqDvpqvCvt6bm8O5d/f+RwXuoRA5w8eVIHDx406+0BAPhTCg0NVe3atS+439TQLSsrU0FBgapVqyaHw2HWZgAA+FNwOp06e/asatasKR+fC4/gmhq6AADgd5xIBQCARQhdAAAsQugCAGARQhcAAIsYFrpbtmxRz549FRsbq0WLFl3wuNPp1NNPP63Y2FjFx8fryy+/NGrTtlVRTz/88EPFx8crPj5egwYNUkZGRhVU6X0q6us5+/fvV1hYmD766CMLq/Ne7vR1586dSkhIUFxcnP76179aXKH3qainJ0+e1JgxY9SnTx/FxcVp5cqVVVCld5kyZYo6d+6s3r17X/Rx07PKaYCSkhLnHXfc4fzhhx+cRUVFzvj4eGdmZma552zevNk5YsQIZ1lZmTMtLc05YMAAIzZtW+70dM+ePc5ff/3V6XT+1l96WjF3+nrueUOHDnWOHDnSuX79+iqo1Lu409f8/HznXXfd5fz555+dTqfTmZubWxWleg13erpgwQLn3LlznU6n03ns2DFnx44dnUVFRVVRrtfYtWuX88CBA864uLiLPm52Vhmy0t2/f79uuOEGNWvWTP7+/oqLi9Nnn31W7jmfffaZ+vbtK4fDofDwcJ04cUI5OTlGbN6W3Olp+/btFRQUJEkKDw/X0aNHq6JUr+JOXyXp7bffVs+ePVW/fv0qqNL7uNPX1atXKzY2Vo0bN5YkelsBd3rqcDhUUFAgp9OpgoICBQUFyc/P1AsNer2OHTu6fm9ejNlZZUjoZmdn67rrrnPdbtiwobKzsy/7nOuuu+6C5+B37vT0fCtWrFDXrl2tKM2rufuzumHDBg0aNMjq8ryWO309dOiQTpw4oaFDh6p///5atWqVxVV6F3d6OmTIEGVlZalLly7q06ePpk6detELMsB9ZmeVIX8SOS9yfY0/XoHKnefgd570a8eOHVqxYoXeffdds8vyeu70ddasWZo8ebJ8fX2tKsvrudPX0tJSffnll1qyZIkKCws1aNAgtWvXTs2bN7eqTK/iTk8///xzhYWFaenSpfrhhx80fPhwRUZGqlatWlaVaTtmZ5UhoXvdddeV27WZnZ2tBg0aXPY5R48eveA5+J07PZWkjIwMTZs2Ta+99prq1atnZYleyZ2+HjhwQA899JAkKS8vTykpKfLz81P37t0trdWbuPs7oF69egoMDFRgYKAiIyOVkZFB6F6COz1NSkrSqFGj5HA4dMMNN6hp06b69ttvdcstt1hdrm2YnVWG7Ie4+eabdejQIf34448qLi7W2rVrFRMTU+45MTExWrVqlZxOp9LT01W7dm1C9zLc6enhw4c1YcIEzZ07l19cbnKnrxs3bnT969mzp6ZPn07gVsCdvt5xxx1KTU1VSUmJzpw5o/379yskJKSKKv7zc6enjRo10vbt2yVJubm5+u6779S0adOqKNc2zM4qQ1a6fn5+euKJJzRy5EiVlpbq7rvvVsuWLfXee+9JkgYPHqxu3bopJSVFsbGxqlGjhmbPnm3Epm3LnZ4mJibq119/1ZNPPilJ8vX1VVJSUlWW/afnTl/hOXf6GhIS4jr26OPjowEDBig0NLSKK//zcqen48aN05QpUxQfHy+n06nJkycrODi4iiv/c3vooYe0a9cu5eXlqWvXrpowYYJKSkokWZNVDDwAAMAinOYGAIBFCF0AACxC6AIAYBFCFwAAixC6AABYhNAFAMAihC4AABYhdAEAsMj/A+yp0iS4WkVLAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 576x396 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from yellowbrick.features import rank1d\n",
    "rank1d(df_aac.to_numpy(), np.where(df_labels.labels == \"Sugar transport\", 1, 0), features = df_aac.columns)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Independent test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    X_train,\n",
    "    X_test,\n",
    "    y_train,\n",
    "    y_test,\n",
    "    sample_names_train,\n",
    "    sample_names_test,\n",
    "    feature_names,\n",
    ") = get_independent_test_set(df_aac, df_labels, test_size=0.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preprocess = make_pipeline(StandardScaler())\n",
    "# X_train_processed = preprocess.fit_transform(X_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model selection\n",
    "\n",
    "SVC (with default RBF kernel) and linear SVM look the most promising. I will choose RBF and switch to linear if the model is overfitting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "###  LinearSVC()\n",
      "CV folds: [0.35  0.35  0.381 0.381 0.381]\n",
      "Mean: 0.369\n",
      "Std: 0.015\n",
      "###  LinearSVC(class_weight='balanced')\n",
      "CV folds: [0.513 0.845 0.606 0.405 0.69 ]\n",
      "Mean: 0.612\n",
      "Std: 0.15\n",
      "###  SVC()\n",
      "CV folds: [0.458 0.35  0.567 0.278 0.567]\n",
      "Mean: 0.444\n",
      "Std: 0.116\n",
      "###  SVC(class_weight='balanced')\n",
      "CV folds: [0.513 0.845 0.675 0.405 0.675]\n",
      "Mean: 0.623\n",
      "Std: 0.151\n",
      "###  GaussianNB()\n",
      "CV folds: [0.513 0.675 0.745 0.513 0.675]\n",
      "Mean: 0.624\n",
      "Std: 0.095\n",
      "###  KNeighborsClassifier()\n",
      "CV folds: [0.575 0.575 0.513 0.511 0.575]\n",
      "Mean: 0.55\n",
      "Std: 0.031\n",
      "###  RandomForestClassifier()\n",
      "CV folds: [0.513 0.675 0.639 0.35  0.458]\n",
      "Mean: 0.527\n",
      "Std: 0.119\n",
      "###  RandomForestClassifier(class_weight='balanced')\n",
      "CV folds: [0.513 0.639 0.639 0.458 0.35 ]\n",
      "Mean: 0.52\n",
      "Std: 0.11\n",
      "###  SGDClassifier()\n",
      "CV folds: [0.536 0.35  0.381 0.278 0.405]\n",
      "Mean: 0.39\n",
      "Std: 0.085\n"
     ]
    }
   ],
   "source": [
    "for estimator in [\n",
    "    LinearSVC(),\n",
    "    LinearSVC(class_weight=\"balanced\"),\n",
    "    SVC(),\n",
    "    SVC(class_weight=\"balanced\"),\n",
    "    GaussianNB(),\n",
    "    KNeighborsClassifier(),\n",
    "    RandomForestClassifier(),\n",
    "    RandomForestClassifier(class_weight=\"balanced\"),\n",
    "    SGDClassifier(),\n",
    "]:\n",
    "    pipeline = make_pipeline(StandardScaler(), estimator)\n",
    "    scores = cross_val_score(estimator, X_train, y_train, scoring=\"f1_macro\")\n",
    "    print(\"### \", estimator)\n",
    "    print(f\"CV folds: {scores.round(3)}\")\n",
    "    print(f\"Mean: {scores.mean().round(3)}\")\n",
    "    print(f\"Std: {scores.std().round(3)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameter tuning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'svc__C': 1, 'svc__class_weight': 'balanced', 'svc__gamma': 0.01}\n",
      "0.6138030760089583\n"
     ]
    }
   ],
   "source": [
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), SVC()),\n",
    "    param_grid={\n",
    "        \"svc__class_weight\": [\"balanced\", None],\n",
    "        \"svc__C\": [0.1, 1, 10],\n",
    "        \"svc__gamma\": [\"scale\", 1e-0, 1e-1, 1e-2],\n",
    "    },\n",
    "    cv=5,\n",
    "    scoring=\"f1_macro\",\n",
    "    n_jobs=-1,\n",
    "    return_train_score=True,\n",
    ")\n",
    "gsearch.fit(X_train, y_train)\n",
    "print(gsearch.best_params_)\n",
    "print(gsearch.best_score_)\n",
    "best_estimator_svc = gsearch.best_estimator_\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dimensionality reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corr</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>corr_abs</th>\n",
       "      <th>mean0</th>\n",
       "      <th>mean1</th>\n",
       "      <th>median0</th>\n",
       "      <th>median1</th>\n",
       "      <th>mediandiff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.319089</td>\n",
       "      <td>0.132892</td>\n",
       "      <td>0.024544</td>\n",
       "      <td>0.319089</td>\n",
       "      <td>0.126342</td>\n",
       "      <td>0.142111</td>\n",
       "      <td>0.126333</td>\n",
       "      <td>0.135084</td>\n",
       "      <td>0.008752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>-0.251708</td>\n",
       "      <td>0.032355</td>\n",
       "      <td>0.010263</td>\n",
       "      <td>0.251708</td>\n",
       "      <td>0.034516</td>\n",
       "      <td>0.029314</td>\n",
       "      <td>0.035646</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.007868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.317947</td>\n",
       "      <td>0.078476</td>\n",
       "      <td>0.015257</td>\n",
       "      <td>0.317947</td>\n",
       "      <td>0.074418</td>\n",
       "      <td>0.084186</td>\n",
       "      <td>0.074494</td>\n",
       "      <td>0.082031</td>\n",
       "      <td>0.007537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.130968</td>\n",
       "      <td>0.038216</td>\n",
       "      <td>0.013223</td>\n",
       "      <td>0.130968</td>\n",
       "      <td>0.036767</td>\n",
       "      <td>0.040254</td>\n",
       "      <td>0.032930</td>\n",
       "      <td>0.040385</td>\n",
       "      <td>0.007455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>-0.218621</td>\n",
       "      <td>0.067955</td>\n",
       "      <td>0.018450</td>\n",
       "      <td>0.218621</td>\n",
       "      <td>0.071329</td>\n",
       "      <td>0.063207</td>\n",
       "      <td>0.072015</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.006798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>-0.154672</td>\n",
       "      <td>0.047714</td>\n",
       "      <td>0.012873</td>\n",
       "      <td>0.154672</td>\n",
       "      <td>0.049380</td>\n",
       "      <td>0.045371</td>\n",
       "      <td>0.047865</td>\n",
       "      <td>0.042308</td>\n",
       "      <td>0.005557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>-0.180204</td>\n",
       "      <td>0.081023</td>\n",
       "      <td>0.013017</td>\n",
       "      <td>0.180204</td>\n",
       "      <td>0.082985</td>\n",
       "      <td>0.078261</td>\n",
       "      <td>0.082572</td>\n",
       "      <td>0.077640</td>\n",
       "      <td>0.004933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>-0.204019</td>\n",
       "      <td>0.027174</td>\n",
       "      <td>0.009984</td>\n",
       "      <td>0.204019</td>\n",
       "      <td>0.028878</td>\n",
       "      <td>0.024776</td>\n",
       "      <td>0.028453</td>\n",
       "      <td>0.024038</td>\n",
       "      <td>0.004414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>-0.071955</td>\n",
       "      <td>0.035430</td>\n",
       "      <td>0.013254</td>\n",
       "      <td>0.071955</td>\n",
       "      <td>0.036228</td>\n",
       "      <td>0.034308</td>\n",
       "      <td>0.036217</td>\n",
       "      <td>0.031936</td>\n",
       "      <td>0.004281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.068783</td>\n",
       "      <td>0.031233</td>\n",
       "      <td>0.008763</td>\n",
       "      <td>0.068783</td>\n",
       "      <td>0.030729</td>\n",
       "      <td>0.031943</td>\n",
       "      <td>0.027525</td>\n",
       "      <td>0.031185</td>\n",
       "      <td>0.003660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>-0.172851</td>\n",
       "      <td>0.029283</td>\n",
       "      <td>0.008418</td>\n",
       "      <td>0.172851</td>\n",
       "      <td>0.030500</td>\n",
       "      <td>0.027570</td>\n",
       "      <td>0.030301</td>\n",
       "      <td>0.027149</td>\n",
       "      <td>0.003152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-0.110893</td>\n",
       "      <td>0.019321</td>\n",
       "      <td>0.006493</td>\n",
       "      <td>0.110893</td>\n",
       "      <td>0.019923</td>\n",
       "      <td>0.018473</td>\n",
       "      <td>0.021333</td>\n",
       "      <td>0.018484</td>\n",
       "      <td>0.002849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.086097</td>\n",
       "      <td>0.075343</td>\n",
       "      <td>0.015141</td>\n",
       "      <td>0.086097</td>\n",
       "      <td>0.074252</td>\n",
       "      <td>0.076877</td>\n",
       "      <td>0.075265</td>\n",
       "      <td>0.072917</td>\n",
       "      <td>0.002349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.059165</td>\n",
       "      <td>0.081131</td>\n",
       "      <td>0.018693</td>\n",
       "      <td>0.059165</td>\n",
       "      <td>0.080206</td>\n",
       "      <td>0.082433</td>\n",
       "      <td>0.078074</td>\n",
       "      <td>0.079840</td>\n",
       "      <td>0.001766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-0.027421</td>\n",
       "      <td>0.040854</td>\n",
       "      <td>0.012212</td>\n",
       "      <td>0.027421</td>\n",
       "      <td>0.041134</td>\n",
       "      <td>0.040460</td>\n",
       "      <td>0.041105</td>\n",
       "      <td>0.039437</td>\n",
       "      <td>0.001668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.018803</td>\n",
       "      <td>0.058895</td>\n",
       "      <td>0.012979</td>\n",
       "      <td>0.018803</td>\n",
       "      <td>0.058691</td>\n",
       "      <td>0.059182</td>\n",
       "      <td>0.058679</td>\n",
       "      <td>0.057199</td>\n",
       "      <td>0.001480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>-0.082713</td>\n",
       "      <td>0.016083</td>\n",
       "      <td>0.007103</td>\n",
       "      <td>0.082713</td>\n",
       "      <td>0.016574</td>\n",
       "      <td>0.015391</td>\n",
       "      <td>0.015983</td>\n",
       "      <td>0.014553</td>\n",
       "      <td>0.001430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-0.038982</td>\n",
       "      <td>0.058241</td>\n",
       "      <td>0.010506</td>\n",
       "      <td>0.038982</td>\n",
       "      <td>0.058584</td>\n",
       "      <td>0.057759</td>\n",
       "      <td>0.057201</td>\n",
       "      <td>0.058161</td>\n",
       "      <td>0.000960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>0.036311</td>\n",
       "      <td>0.034650</td>\n",
       "      <td>0.010153</td>\n",
       "      <td>0.036311</td>\n",
       "      <td>0.034342</td>\n",
       "      <td>0.035084</td>\n",
       "      <td>0.033955</td>\n",
       "      <td>0.034527</td>\n",
       "      <td>0.000571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>-0.092618</td>\n",
       "      <td>0.013730</td>\n",
       "      <td>0.006353</td>\n",
       "      <td>0.092618</td>\n",
       "      <td>0.014222</td>\n",
       "      <td>0.013037</td>\n",
       "      <td>0.012831</td>\n",
       "      <td>0.013333</td>\n",
       "      <td>0.000502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       corr      mean       std  corr_abs     mean0     mean1   median0  \\\n",
       "L  0.319089  0.132892  0.024544  0.319089  0.126342  0.142111  0.126333   \n",
       "N -0.251708  0.032355  0.010263  0.251708  0.034516  0.029314  0.035646   \n",
       "G  0.317947  0.078476  0.015257  0.317947  0.074418  0.084186  0.074494   \n",
       "R  0.130968  0.038216  0.013223  0.130968  0.036767  0.040254  0.032930   \n",
       "I -0.218621  0.067955  0.018450  0.218621  0.071329  0.063207  0.072015   \n",
       "P -0.154672  0.047714  0.012873  0.154672  0.049380  0.045371  0.047865   \n",
       "V -0.180204  0.081023  0.013017  0.180204  0.082985  0.078261  0.082572   \n",
       "D -0.204019  0.027174  0.009984  0.204019  0.028878  0.024776  0.028453   \n",
       "K -0.071955  0.035430  0.013254  0.071955  0.036228  0.034308  0.036217   \n",
       "Q  0.068783  0.031233  0.008763  0.068783  0.030729  0.031943  0.027525   \n",
       "M -0.172851  0.029283  0.008418  0.172851  0.030500  0.027570  0.030301   \n",
       "C -0.110893  0.019321  0.006493  0.110893  0.019923  0.018473  0.021333   \n",
       "S  0.086097  0.075343  0.015141  0.086097  0.074252  0.076877  0.075265   \n",
       "A  0.059165  0.081131  0.018693  0.059165  0.080206  0.082433  0.078074   \n",
       "E -0.027421  0.040854  0.012212  0.027421  0.041134  0.040460  0.041105   \n",
       "F  0.018803  0.058895  0.012979  0.018803  0.058691  0.059182  0.058679   \n",
       "W -0.082713  0.016083  0.007103  0.082713  0.016574  0.015391  0.015983   \n",
       "T -0.038982  0.058241  0.010506  0.038982  0.058584  0.057759  0.057201   \n",
       "Y  0.036311  0.034650  0.010153  0.036311  0.034342  0.035084  0.033955   \n",
       "H -0.092618  0.013730  0.006353  0.092618  0.014222  0.013037  0.012831   \n",
       "\n",
       "    median1  mediandiff  \n",
       "L  0.135084    0.008752  \n",
       "N  0.027778    0.007868  \n",
       "G  0.082031    0.007537  \n",
       "R  0.040385    0.007455  \n",
       "I  0.065217    0.006798  \n",
       "P  0.042308    0.005557  \n",
       "V  0.077640    0.004933  \n",
       "D  0.024038    0.004414  \n",
       "K  0.031936    0.004281  \n",
       "Q  0.031185    0.003660  \n",
       "M  0.027149    0.003152  \n",
       "C  0.018484    0.002849  \n",
       "S  0.072917    0.002349  \n",
       "A  0.079840    0.001766  \n",
       "E  0.039437    0.001668  \n",
       "F  0.057199    0.001480  \n",
       "W  0.014553    0.001430  \n",
       "T  0.058161    0.000960  \n",
       "Y  0.034527    0.000571  \n",
       "H  0.013333    0.000502  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train = pd.DataFrame(X_train, columns=feature_names, index=sample_names_train)\n",
    "get_feature_stats(df_train, df_labels.loc[df_train.index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'selectkbest__k': 14, 'svc__C': 1, 'svc__class_weight': 'balanced', 'svc__gamma': 'scale'}\n",
      "0.671\n",
      "Removed features  ['A' 'E' 'F' 'Q' 'T' 'Y']\n"
     ]
    }
   ],
   "source": [
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\"],\n",
    "    \"svc__C\": [1, 10, 100],\n",
    "    \"svc__gamma\": [\"scale\", 1e-1, 1e-2, 1e-3],\n",
    "    \"selectkbest__k\": list(range(1, 20)),\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), SelectKBest(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=5,\n",
    "    scoring=\"f1_macro\",\n",
    "    n_jobs=-1,\n",
    "    return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "gsearch.fit(X_train, y_train)\n",
    "print(gsearch.best_params_)\n",
    "print(gsearch.best_score_.round(3))\n",
    "print(\n",
    "    \"Removed features \",\n",
    "    feature_names[~gsearch.best_estimator_[\"selectkbest\"].get_support()],\n",
    ")\n",
    "best_estimator_svc_kbest = gsearch.best_estimator_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'pca__n_components': 0.8200000000000001, 'svc__C': 1, 'svc__class_weight': 'balanced', 'svc__gamma': 'scale'}\n",
      "0.680058275058275\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contrib. PC1</th>\n",
       "      <th>Contrib. PC2</th>\n",
       "      <th>Contrib. PC3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.292196</td>\n",
       "      <td>0.075781</td>\n",
       "      <td>-0.205078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.285977</td>\n",
       "      <td>0.20073</td>\n",
       "      <td>-0.233172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.276812</td>\n",
       "      <td>-0.19994</td>\n",
       "      <td>-0.221125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>0.267199</td>\n",
       "      <td>0.114004</td>\n",
       "      <td>0.373787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.222661</td>\n",
       "      <td>0.368936</td>\n",
       "      <td>0.283644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.165814</td>\n",
       "      <td>-0.2558</td>\n",
       "      <td>0.096281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>0.15267</td>\n",
       "      <td>-0.242973</td>\n",
       "      <td>0.17874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.079748</td>\n",
       "      <td>-0.348014</td>\n",
       "      <td>0.017727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.06847</td>\n",
       "      <td>0.213984</td>\n",
       "      <td>0.051695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.006812</td>\n",
       "      <td>-0.054565</td>\n",
       "      <td>0.487303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>-0.07127</td>\n",
       "      <td>-0.358209</td>\n",
       "      <td>0.028532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>-0.099772</td>\n",
       "      <td>0.144004</td>\n",
       "      <td>-0.42311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>-0.11971</td>\n",
       "      <td>0.063524</td>\n",
       "      <td>0.314946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-0.16327</td>\n",
       "      <td>0.342424</td>\n",
       "      <td>0.045928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-0.21506</td>\n",
       "      <td>-0.318855</td>\n",
       "      <td>0.074069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>-0.219688</td>\n",
       "      <td>0.270589</td>\n",
       "      <td>0.10534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-0.242475</td>\n",
       "      <td>0.105053</td>\n",
       "      <td>-0.105635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>-0.323855</td>\n",
       "      <td>-0.034231</td>\n",
       "      <td>0.178437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>-0.336429</td>\n",
       "      <td>-0.101329</td>\n",
       "      <td>-0.102976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>-0.382031</td>\n",
       "      <td>0.053159</td>\n",
       "      <td>-0.038944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Contrib. PC1 Contrib. PC2 Contrib. PC3\n",
       "Feature                                       \n",
       "G           0.292196     0.075781    -0.205078\n",
       "A           0.285977      0.20073    -0.233172\n",
       "L           0.276812     -0.19994    -0.221125\n",
       "P           0.267199     0.114004     0.373787\n",
       "R           0.222661     0.368936     0.283644\n",
       "C           0.165814      -0.2558     0.096281\n",
       "W            0.15267    -0.242973      0.17874\n",
       "S           0.079748    -0.348014     0.017727\n",
       "Q            0.06847     0.213984     0.051695\n",
       "H           0.006812    -0.054565     0.487303\n",
       "Y           -0.07127    -0.358209     0.028532\n",
       "V          -0.099772     0.144004     -0.42311\n",
       "D           -0.11971     0.063524     0.314946\n",
       "E           -0.16327     0.342424     0.045928\n",
       "F           -0.21506    -0.318855     0.074069\n",
       "M          -0.219688     0.270589      0.10534\n",
       "T          -0.242475     0.105053    -0.105635\n",
       "N          -0.323855    -0.034231     0.178437\n",
       "I          -0.336429    -0.101329    -0.102976\n",
       "K          -0.382031     0.053159    -0.038944"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of components to explain variance: 12\n"
     ]
    }
   ],
   "source": [
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\", None],\n",
    "    \"svc__C\": [0.1, 1, 10],\n",
    "    \"svc__gamma\": [\"scale\",1e-1, 1e-2, 1e-3],\n",
    "    \"pca__n_components\": np.linspace(0.8, 0.99, 20),\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), PCA(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=5,\n",
    "    scoring=\"f1_macro\",\n",
    "    n_jobs=-1,\n",
    "    return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "gsearch.fit(X_train, y_train)\n",
    "print(gsearch.best_params_)\n",
    "print(gsearch.best_score_)\n",
    "best_estimator_svc_pca = gsearch.best_estimator_\n",
    "display(\n",
    "    pd.DataFrame.from_records(\n",
    "        [\n",
    "            feature_names,\n",
    "            gsearch.best_estimator_[\"pca\"].components_[0],\n",
    "            gsearch.best_estimator_[\"pca\"].components_[1],\n",
    "            gsearch.best_estimator_[\"pca\"].components_[2],\n",
    "        ],\n",
    "        index=[\"Feature\", \"Contrib. PC1\", \"Contrib. PC2\", \"Contrib. PC3\"],\n",
    "    )\n",
    "    .T.set_index(\"Feature\")\n",
    "    .sort_values(\"Contrib. PC1\", ascending=False)\n",
    ")\n",
    "\n",
    "pca = PCA()\n",
    "pca.fit(X_train)\n",
    "csum = np.cumsum(pca.explained_variance_ratio_)\n",
    "print(\"Number of components to explain variance:\", np.argmax(csum >= 0.93) + 1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train scores: 0.671+-0.06\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino</th>\n",
       "      <th>Sugar</th>\n",
       "      <th>Macro</th>\n",
       "      <th>Weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>precision</th>\n",
       "      <td>0.667</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.583</td>\n",
       "      <td>0.598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>recall</th>\n",
       "      <td>0.600</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.586</td>\n",
       "      <td>0.588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f1-score</th>\n",
       "      <td>0.632</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.582</td>\n",
       "      <td>0.591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>support</th>\n",
       "      <td>10.000</td>\n",
       "      <td>7.000</td>\n",
       "      <td>17.000</td>\n",
       "      <td>17.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Amino  Sugar   Macro  Weighted\n",
       "precision   0.667  0.500   0.583     0.598\n",
       "recall      0.600  0.571   0.586     0.588\n",
       "f1-score    0.632  0.533   0.582     0.591\n",
       "support    10.000  7.000  17.000    17.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino</th>\n",
       "      <th>Sugar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Amino</th>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sugar</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Amino  Sugar\n",
       "Amino      6      4\n",
       "Sugar      3      4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "best_estimator = best_estimator_svc_kbest\n",
    "best_scores = cross_val_score(\n",
    "    estimator=clone(best_estimator), X=X_train, y=y_train, scoring=\"f1_macro\"\n",
    ")\n",
    "print(f\"Train scores: {best_scores.mean().round(3)}+-{best_scores.std().round(3)}\")\n",
    "\n",
    "y_pred = best_estimator.predict(X_test)\n",
    "y_true = y_test.copy()\n",
    "\n",
    "report_df, confusion_matrix_df = print_validation_results(y_true, y_pred, labels=[\"Amino\", \"Sugar\"])\n",
    "display(report_df.round(3))\n",
    "display(confusion_matrix_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train scores: 0.68+-0.115\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino</th>\n",
       "      <th>Sugar</th>\n",
       "      <th>Macro</th>\n",
       "      <th>Weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>precision</th>\n",
       "      <td>0.636</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.568</td>\n",
       "      <td>0.580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>recall</th>\n",
       "      <td>0.700</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.564</td>\n",
       "      <td>0.588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f1-score</th>\n",
       "      <td>0.667</td>\n",
       "      <td>0.462</td>\n",
       "      <td>0.564</td>\n",
       "      <td>0.582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>support</th>\n",
       "      <td>10.000</td>\n",
       "      <td>7.000</td>\n",
       "      <td>17.000</td>\n",
       "      <td>17.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Amino  Sugar   Macro  Weighted\n",
       "precision   0.636  0.500   0.568     0.580\n",
       "recall      0.700  0.429   0.564     0.588\n",
       "f1-score    0.667  0.462   0.564     0.582\n",
       "support    10.000  7.000  17.000    17.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino</th>\n",
       "      <th>Sugar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Amino</th>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sugar</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Amino  Sugar\n",
       "Amino      7      3\n",
       "Sugar      4      3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "best_estimator = best_estimator_svc_pca\n",
    "best_scores = cross_val_score(\n",
    "    estimator=clone(best_estimator), X=X_train, y=y_train, scoring=\"f1_macro\"\n",
    ")\n",
    "print(f\"Train scores: {best_scores.mean().round(3)}+-{best_scores.std().round(3)}\")\n",
    "\n",
    "y_pred = best_estimator.predict(X_test)\n",
    "y_true = y_test.copy()\n",
    "\n",
    "report_df, confusion_matrix_df = print_validation_results(y_true, y_pred, labels=[\"Amino\", \"Sugar\"])\n",
    "display(report_df.round(3))\n",
    "display(confusion_matrix_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## With nested LOOCV\n",
    "\n",
    "There are not enough samples in dataset, trying different eval method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6707317073170732\n",
      "0.6829268292682927\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import LeaveOneOut\n",
    "\n",
    "X = df_aac.values\n",
    "labels = df_labels.labels\n",
    "y = np.where(labels == \"Sugar transport\", 1, 0)\n",
    "\n",
    "\n",
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\"],\n",
    "    \"svc__C\": [0.1, 1, 10],\n",
    "    \"svc__gamma\": [\"scale\", 1e-1, 1e-2, 1e-3],\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=LeaveOneOut(),\n",
    "    scoring=\"f1_macro\",\n",
    "    # n_jobs=1,\n",
    "    return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "# the score does not make a difference, since it's either 1 (correct) or 0 (incorrect)\n",
    "res = cross_val_score(gsearch, X, y, scoring=\"f1_macro\", n_jobs=-1, cv=LeaveOneOut())\n",
    "print(res.mean())\n",
    "\n",
    "\n",
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\"],\n",
    "    \"svc__C\": [0.1, 1, 10],\n",
    "    \"svc__gamma\": [\"scale\", 1e-1, 1e-2, 1e-3],\n",
    "    \"pca__n_components\": [0.8, 0.9],\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), PCA(), StandardScaler(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=LeaveOneOut(),\n",
    "    scoring=\"f1_macro\",\n",
    "    # n_jobs=1,\n",
    "    return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "# the score does not make a difference, since it's either 1 (correct) or 0 (incorrect)\n",
    "res = cross_val_score(gsearch, X, y, scoring=\"f1_macro\", n_jobs=-1, cv=LeaveOneOut())\n",
    "print(res.mean())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6667168447990366\n",
      "0.6872616323417239\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import LeaveOneOut\n",
    "\n",
    "X = df_aac.values\n",
    "labels = df_labels.labels\n",
    "y = np.where(labels == \"Sugar transport\", 1, 0)\n",
    "\n",
    "\n",
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\"],\n",
    "    \"svc__C\": [0.1, 1, 10],\n",
    "    \"svc__gamma\": [\"scale\", 1e-1, 1e-2, 1e-3],\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=LeaveOneOut(),\n",
    "    scoring=\"f1_macro\",\n",
    "    # n_jobs=1,\n",
    "    return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "# the score does not make a difference, since it's either 1 (correct) or 0 (incorrect)\n",
    "res = cross_val_predict(gsearch, X, y, n_jobs=-1, cv=LeaveOneOut())\n",
    "print(f1_score(y, res, average=\"macro\"))\n",
    "\n",
    "\n",
    "params_svc = {\n",
    "    \"svc__class_weight\": [\"balanced\"],\n",
    "    \"svc__C\": [0.1, 1, 10],\n",
    "    # \"svc__gamma\": [\"scale\", 1e-1, 1e-2, 1e-3],\n",
    "    \"pca__n_components\": [0.8, 0.9],\n",
    "}\n",
    "gsearch = GridSearchCV(\n",
    "    make_pipeline(StandardScaler(), PCA(), StandardScaler(), SVC()),\n",
    "    param_grid=params_svc,\n",
    "    cv=LeaveOneOut(),\n",
    "    scoring=\"f1_macro\",\n",
    "    # n_jobs=1,\n",
    "    # return_train_score=True,\n",
    "    refit=True,\n",
    ")\n",
    "# the score does not make a difference, since it's either 1 (correct) or 0 (incorrect)\n",
    "res = cross_val_predict(gsearch, X, y, n_jobs=-1, cv=LeaveOneOut())\n",
    "print(f1_score(y, res, average=\"macro\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('subpred')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7d9587acfc26b94842e9ed1391c37acbb1c055ffb7f48359e727081012dc2cc7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
